server/src/main/java/org/elasticsearch/reservedstate/service/FileSettingsService.java

```diff
@@ -0,0 +1,330 @@
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.reservedstate.service;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.cluster.ClusterChangedEvent;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.ClusterStateListener;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.component.AbstractLifecycleComponent;
+import org.elasticsearch.common.settings.Setting;
+import org.elasticsearch.core.PathUtils;
+import org.elasticsearch.env.Environment;
+import org.elasticsearch.xcontent.XContentParser;
+import org.elasticsearch.xcontent.XContentParserConfiguration;
+
+import java.io.BufferedInputStream;
+import java.io.IOException;
+import java.nio.file.ClosedWatchServiceException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.StandardWatchEventKinds;
+import java.nio.file.WatchKey;
+import java.nio.file.WatchService;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.util.concurrent.CountDownLatch;
+
+import static org.elasticsearch.xcontent.XContentType.JSON;
+
+/**
+ * File based settings applier service which watches an 'operator` directory inside the config directory.
+ * <p>
+ * The service expects that the operator directory will contain a single JSON file with all the settings that
+ * need to be applied to the cluster state. The name of the file is fixed to be settings.json. The operator
+ * directory name can be configured by setting the 'path.config.operator_directory' in the node properties.
+ * <p>
+ * The {@link FileSettingsService} is active always, but enabled only on the current master node. We register
+ * the service as a listener to cluster state changes, so that we can enable the file watcher thread when this
+ * node becomes a master node.
+ */
+public class FileSettingsService extends AbstractLifecycleComponent implements ClusterStateListener {
+    private static final Logger logger = LogManager.getLogger(FileSettingsService.class);
+
+    private static final String SETTINGS_FILE_NAME = "settings.json";
+    static final String NAMESPACE = "file_settings";
+
+    private final ReservedClusterStateService stateService;
+    private final Path operatorSettingsDir;
+
+    private WatchService watchService; // null;
+    private CountDownLatch watcherThreadLatch;
+
+    private volatile FileUpdateState fileUpdateState = null;
+    private volatile WatchKey settingsDirWatchKey = null;
+
+    private volatile boolean active = false;
+    private volatile boolean initialState = true;
+
+    public static final Setting<String> OPERATOR_DIRECTORY = Setting.simpleString(
```

Comment:
Why is this configurable? I think we could start with it hardcoded? I much prefer it is always the same, so that we know it will always be directly under the config dir.

Reply:
Sure thing, I'll fix it.